app/src/production/java/com/marvel/marvel/main/dagger/MainModule.kt

```diff
@@ -0,0 +1,56 @@
+package com.marvel.marvel.main.dagger
+
+
+import android.arch.lifecycle.ViewModel
+import android.arch.lifecycle.ViewModelProvider
+import android.arch.lifecycle.ViewModelProviders
+import android.support.v7.app.AppCompatActivity
+import com.marvel.marvel.application.PicassoImageLoader
+import com.marvel.marvel.main.model.Api
+import com.marvel.marvel.main.model.ApiRetrofit
+import com.marvel.marvel.main.model.MainModel
+import com.marvel.marvel.main.model.MainModelRetrofit
+import com.marvel.marvel.main.presenter.MainPresenter
+import dagger.Module
+import dagger.Provides
+import rx.android.schedulers.AndroidSchedulers
+import rx.schedulers.Schedulers
+
+
+@Module
+class MainModule(private val activity: AppCompatActivity) {
+
+    @SuppressWarnings("unused")
+    @Provides
+    internal fun provideMainPresenter(model: MainModel) =
+        MainPresenter(model, activity.applicationContext)
+
+    @SuppressWarnings("unused")
+    @Provides
+    internal fun provideApi(): Api = ApiRetrofit()
+
+    @SuppressWarnings("unused")
+    @Provides
+    internal fun provideMainModel(api: Api, factory: ViewModelProvider.Factory): MainModel =
+        getModel(factory)
+
+    @SuppressWarnings("unused")
+    @Provides
+    internal fun provideLoader() = PicassoImageLoader
+
+    @SuppressWarnings("unused")
+    @Provides
+    internal fun provideFactory(api: Api): ViewModelProvider.Factory = RetrofitViewModelFactory(api)
+
+    private fun getModel(factory: ViewModelProvider.Factory): MainModel =
+        ViewModelProviders.of(activity, factory).get(MainModelRetrofit::class.java)
+
+    class RetrofitViewModelFactory(val api: Api) : ViewModelProvider.Factory {
+
+        @Suppress("UNCHECKED_CAST")
+        override fun <T : ViewModel?> create(p0: Class<T>): T =
+            MainModelRetrofit(api, Schedulers.newThread(), AndroidSchedulers.mainThread()) as T
```

Comment:
You almost never want to `Schedulers.newThread()`.

`Schedulers.IO()` is the suitable `Scheduler` here.

Also with retrofit you can tell the whole system to use the correct scheduler:

```
Retrofit.Builder()
                // Note: this line puts all Rx calls on the IO scheduler, so no need to specify subscribeOn later
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createWithScheduler(Schedulers.io()))
```

I also, do not pass `AndroidSchedulers.mainThread()` around. That's the kind of cross-cutting concern that I'm happy to not inject.

Reply:
Good point about using `io()` and passing the scheduler directly to Retrofit.

Why wouldn't you inject the main thread scheduler? Can't it be useful when you test (because you could inject another scheduler like `trampoline`)?

Reply:
Good question, fortunately, it is easy to test with a rule that does this:
```
RxAndroidPlugins.setMainThreadSchedulerHandler { Schedulers.trampoline() }
```

Reset:

```
RxAndroidPlugins.reset()
```

RxJava has the same for each of its schedulers:

```
RxJavaPlugins.setIoSchedulerHandler { Schedulers.trampoline() }
```

Reset:

```
RxJavaPlugins.reset()
```